### PR TITLE
Implemented basic Quality Attenuation

### PR DESCRIPTION
**Pull request by bjornite**

### CODE REVIEW

File: qualityattenuation/qualityattenuation.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+package qualityattenuation
+
+import (
+	"math"
+
+	"github.com/influxdata/tdigest"
+)
+
+type SimpleQualityAttenuation struct {
+	EmpiricalDistribution     *tdigest.TDigest
```
Comment (hawkinsw) [inline qualityattenuation/qualityattenuation.go:12]: In go when you start a member name with a capital letter that makes it public. Because you have done the great work of making nice *getters*, I don't think that we need to have these as public. 

(see below)

File: qualityattenuation/qualityattenuation.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	}
+}
+
+func (qa *SimpleQualityAttenuation) AddSample(sample float64) {
+	if sample <= 0.0 {
+		// Ignore zero or negative samples because they cannot be valid.
+		// TODO: This should raise a warning and/or trigger error handling.
+		return
+	}
+	qa.Number_of_samples++
```
Comment (hawkinsw) [inline qualityattenuation/qualityattenuation.go:43]: In go, when you implement the member function you can have access to the private fields of the struct. 

TL;DR: You are using *excellent* go coding without even knowing it!!

File: qualityattenuation/qualityattenuation.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	"math"
+
+	"github.com/influxdata/tdigest"
+)
+
+type SimpleQualityAttenuation struct {
+	EmpiricalDistribution     *tdigest.TDigest
+	Offset                    float64
+	Offset_sum                float64
+	Offset_sum_of_squares     float64
```
Comment (hawkinsw) [inline qualityattenuation/qualityattenuation.go:15]: If you don't mind, can we convert to `camelCase` (what go calls `MixedCase`? I prefer `snake_case` but go habits are for the former. (https://go.dev/doc/effective_go#mixed-caps)

File: qualityattenuation/qualityattenuation.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+		Latency_eq_loss_threshold: 15.0, // Count latency greater than this value as a loss.
+		Minimum_latency:           0.0,
+		Maximum_latency:           0.0,
+	}
+}
+
+func (qa *SimpleQualityAttenuation) AddSample(sample float64) {
+	if sample <= 0.0 {
+		// Ignore zero or negative samples because they cannot be valid.
+		// TODO: This should raise a warning and/or trigger error handling.
```
Comment (hawkinsw) [inline qualityattenuation/qualityattenuation.go:40]: Is this case something that the caller would be able to do something about? If it's not, then we can just eat up the error silently and move on. Otherwise, we could make the method return an `error`. If the function succeeds, we return `nil` and otherwise return the result of `fmt.Errorf` (or something like that) which the caller can detect and change their behavior. Again, though, if the caller cannot change their behavior and would simply accept that the datapoint could not be added, perhaps there is no reason to return anything? 

On second thought, even if the caller can only log that an error occurred, perhaps returning an error is a "good idea"? 

I am curious what you think!

File: qualityattenuation/qualityattenuation.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	return qa.GetPercentile(percentile) - qa.GetMinimum()
+}
+
+// Merge two quality attenuation values. This operation assumes the two samples have the same offset and latency_eq_loss_threshold, and
+// will return an error if they do not.
+// It also assumes that the two quality attenuation values are measurements of the same thing (path, outcome, etc.).
+func (qa *SimpleQualityAttenuation) Merge(other *SimpleQualityAttenuation) {
+	// Check that offsets are the same
+	if qa.Offset != other.Offset || qa.Latency_eq_loss_threshold != other.Latency_eq_loss_threshold {
+		//"Cannot merge quality attenuation values with different offset or latency_eq_loss_threshold"
```
Comment (hawkinsw) [inline qualityattenuation/qualityattenuation.go:116]: We can have the method return an `error` and you can generate one with this excellent message by calling `fmt.Errorf`.  Again, go convention is that `error` descriptions do not start with uppercase letters -- odd, I know!!

File: networkQuality.go [hunk 6]
```diff
@@ -543,6 +550,7 @@ timeout:
 					}
 				} else if probeMeasurement.Type == probe.SelfDown || probeMeasurement.Type == probe.SelfUp {
 					selfRtts.AddElement(probeMeasurement.Duration.Seconds())
+					selfRttsQualityAttenuation.AddSample(probeMeasurement.Duration.Seconds())
```
Comment (hawkinsw) [inline networkQuality.go:553]: Is adding a sample compute intensive? If so, perhaps we only want to add it when the user has the functionality enabled?

**Review by hawkinsw — COMMENTED** [review]
Really, really awesome!! Thank you so much!

File: qualityattenuation/qualityattenuation.go (hunk 4 above)
Comment (bjornite) [reply qualityattenuation/qualityattenuation.go:40]: I added the return value as you proposed.  I don't have a very strong opinion either way on this one, but it seems useful that the caller can be  notified when a sample is dropped.

File: networkQuality.go (hunk 6 above)
Comment (bjornite) [reply networkQuality.go:553]: Not very, at least not with the default level of compression I've used. In the tests I add 160000 samples and the tests run in ~20 milliseconds on my machine. Checking if it's necessary seems like a good idea either way though, so I've added that.

**Review by hawkinsw — APPROVED** [review]
Thanks!!
